### ISSUE TITLE
Felix is thrashing CPU and increasing etcd usage at 60s intervals

### ISSUE THREAD

**Issue by alexwlchan** [Metaswitch/calico#615]
In a recent simulated scale test, I noticed that every 60 seconds:

* Felix has a spike of 80% CPU 
* etcd has a spike of 20% CPU
* RAM disk usage jumps by ~120 kiB

This occurs even when the test framework is idle. Full results are [on Google Drive](https://docs.google.com/document/d/1w_r8SceJA7bJ9NOD4bmgqvnOHg0bny7rYF9qHG4vwDk); the relevant graphs start on page 3.

I saw similar results in the [previous simulated scale tests](https://docs.google.com/document/d/1Se-HGlb28CCy9tCGsJFF-qymHwAjqHXQnVVq_NyN1O4) (see page 8), but we mostly overlooked them because a 60s interval was too short to notice in the other tests.

It seems likely that something is going wrong in the periodic resync (which happens on a 60s interval). What's unusual is that the periodic resync code comes from mech_calico.py, so it's not obvious why that would affect Felix.

For a short-term fix, we can bump up the resync interval, but we should really find out what's leaking, and fix it.

**Comment by fasaxc** [comment]
etcd is garbage collected so it's normal operation is to increase disk usage up to some limit and then do a GC, which should reduce it back down again.  It's important to run the test long enough to get it to do a GC.  It'll take a long time without heavy churn going on, I think we have to make several thousand writes before it GCs.

The CPU spikes at the pluign/etcd are normal, I think.  The periodic resync is intensive and should use plenty of CPU if available.

However, if the periodic resync doesn't find anything to change then I'd expect it not to update etcd, so felix shouldn't see any changes and hence it shouldn't do anything.

**Comment by fasaxc** [comment]
@alexwlchan are there any logs from the periodic resync on the control node?  I think it should log if it finds anything that it needs to do.

**Comment by fasaxc** [comment]
(Or from felix, for that matter.)

**Comment by fasaxc** [comment]
@plwhite I think @alexwlchan is on vacation, are we likely to get any more diags or is anyone able to repro?

**Comment by plwhite** [comment]
I can repro easily enough - just set up a system with a lot of data in it (a few thousand endpoints, a thousand hosts) and wait a bit. You'll see spikes of CPU every 60 seconds.

This is caused by the OpenStack plugin resyncing its data (we think). Most likely it's a simple bug where the data read back is not as expected and so it rewrites something.

If you want to investigate, feel free to take the issue; otherwise I'll get to it later in the week.

**Comment by Lukasa** [comment]
That's almost certainly right.

OpenStack resync is something we need to continue to think about over the next few weeks to try to reduce its impact. I'm going to sketch out some possible enhancements and solicit feedback from the team, I think.

**Comment by alexwlchan** [comment]
I've done a bit more testing, and discovered that I can reproduce this bug on the newest version of Calico (0.25). I've also captured a stack of debug logs from Felix, which we can go through to hopefully narrow down the bug. Details [on Google Drive](https://docs.google.com/document/d/1pow7C395OggXN5kSv_3yplq5BcIziCXqYqcxvvY2Rd4/edit).

**Comment by plwhite** [comment]
Here's a relevant part of the logs. Looks like we fail to correctly parse a timeout of some sort.

    2015-06-22 14:00:50,676 [DEBUG][46626/8] calico.felix.actor 245: Finishing message batch
    2015-06-22 14:01:52,051 [ERROR][46626/3] calico.felix.fetcd 371: Unknown etcd error 'Unable to decode server response: timed out'; doing resync.
    Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/calico/felix/fetcd.py", line 329, in _wait_for_etcd_event
        check_cluster_uuid=True)
      File "/usr/local/lib/python2.7/dist-packages/python_etcd-0.3.3_calico_2-py2.7.egg/etcd/client.py", line 357, in read
        return self._result_from_response(response)
      File "/usr/local/lib/python2.7/dist-packages/python_etcd-0.3.3_calico_2-py2.7.egg/etcd/client.py", line 541, in _result_from_response
        'Unable to decode server response: %s' % e)
    EtcdException: Unable to decode server response: timed out
    2015-06-22 14:01:52,051 [ERROR][46626/3] calico.felix.fetcd 371: Unknown etcd error 'Unable to decode server response: timed out'; doing resync.
    Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/calico/felix/fetcd.py", line 329, in _wait_for_etcd_event
        check_cluster_uuid=True)
      File "/usr/local/lib/python2.7/dist-packages/python_etcd-0.3.3_calico_2-py2.7.egg/etcd/client.py", line 357, in read
        return self._result_from_response(response)
      File "/usr/local/lib/python2.7/dist-packages/python_etcd-0.3.3_calico_2-py2.7.egg/etcd/client.py", line 541, in _result_from_response
        'Unable to decode server response: %s' % e)
    EtcdException: Unable to decode server response: timed out
    2015-06-22 14:01:52,052 [INFO][46626/3] calico.felix.fetcd 199: (Re)connecting to etcd. Old etcd cluster ID was f2e800cca50551bc.
    2015-06-22 14:01:52,054 [INFO][46626/3] calico.felix.fetcd 231: Polling aborted, doing resync.
    2015-06-22 14:01:52,054 [INFO][46626/3] calico.felix.fetcd 216: Reconnecting and loading snapshot from etcd...
    2015-06-22 14:01:52,055 [INFO][46626/3] calico.felix.fetcd 201: (Re)connecting to etcd. No previous cluster ID.

**Comment by fasaxc** [comment]
I think we fixed this already with an update to our fork of python-etcd: https://github.com/Metaswitch/python-etcd/commit/03ea4610e6bf22e28ff5830953f47f4ea4664efb

The problem was that the timeout was being re-wrapped instead of being propagated.

**Comment by plwhite** [comment]
Good stuff - so follow on questions.

- This goes away as soon as we take the updated `python-etcd`, i.e. next release (or a release soon), right? Or does something have to happen on that one.

- Are we getting these fixes upstreamed into python-etcd (assuming we are, but not happened yet)?

**Comment by alexwlchan** [comment]
I have the latest version of etcd on the scale test rig, but I don't know what version of python-etcd it's using. I'll have a look tomorrow and see if it needs updating. (If so, it wasn't updated by the dpkg's – is that a problem?)

**Comment by plwhite** [comment]
I suspect that the version of `python-etcd` in our PPAs is out of date, and hasn't got this fix yet. We could get the version of `python-etcd` direct from the GitHub repo and put that in - might be worth trying if it's quick.

**Comment by Lukasa** [comment]
I don't *think* that's true, I recall checking yesterday and confirming that it wasn't the case.

**Comment by alexwlchan** [comment]
So here's the version of python-etcd on my test kit:

    $ dpkg -l | grep etcd
    ii  etcd             2.0.11-1~ubuntu14.04.1~ppa3      amd64        Highly-available key value
                                                                       store for shared configuration
                                                                       and service discovery
    ii  python-etcd      0.3.3-1calico0.22.1              all          A python client for etcd

I updated from the 0.22.1 release to the 0.25 release yesterday, but didn't touch python-etcd. Since Shaun's fix went in to python-etcd after Calico 0.22.1 was released, I assume that's why I don't have the fix yet. I'll update to the latest version of python-etcd to check if this goes away.

**Comment by Lukasa** [comment]
I think we released python-etcd mid-cycle, and gave it the release number of the active release at the time. I don't think we missed this.

**Comment by alexwlchan** [comment]
1. I've just tried to update python-etcd, and apparently it can't find any updates:

   <pre><code><strong>root at calico23 in /h/ubuntu
   # dpkg -l | grep etcd</strong>
   ii  etcd             2.0.11-1~ubuntu14.04.1~ppa3      amd64        Highly-available key value store for shared configuration and service discovery
   ii  python-etcd      0.3.3-1calico0.22.1              all          A python client for etcd

    <strong>root at calico23 in /h/ubuntu
    # apt-get install python-etcd</strong>
    Reading package lists... Done
    Building dependency tree
    Reading state information... Done
    python-etcd is already the newest version.
    0 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.</code></pre>

    I also don't see a [release of python-etcd](https://github.com/Metaswitch/python-etcd/releases) from our fork that comes after Shaun's fix.

2. I'm going to take the latest code from our python-etcd fork and repeat the test. If it does, we should probably think about updating our PPAs so that other people don't get bitten by this bug.

**Comment by Lukasa** [comment]
I think what I did was check whether code from Shaun's patch was in the released package, which I believe it was. But maybe I hallucinated the entire procedure, it wouldn't be unheard of.

**Comment by alexwlchan** [comment]
I've repeated the test with an updated version of python-etcd. It has no effect on etcd usage (etcd's virtual memory usage is somewhat higher, but that seems unrelated). Updated graphs and logs are in Google Drive.

I had another look through the logs: I can't find anything that looks like the etcd timeout message above, so that bug seems to have been squashed. But I did see this log fragment which seems to suggest a lot of etcd work is being done:

```
110942:2015-06-23 11:08:22,105 [DEBUG][39638/3] urllib3.connectionpool 390: Setting read timeout to 90
110943:2015-06-23 11:08:22,106 [DEBUG][39638/3] urllib3.connectionpool 430: "GET /v2/keys/calico/v1?waitIndex=2291&recursive=true&wait=true HTTP/1.1" 200 None
110944:2015-06-23 11:08:22,106 [DEBUG][39638/3] calico.felix.fetcd 334: etcd response: <class 'etcd.EtcdResult'>({'newKey': False, '_prev_node': <class 'etcd.EtcdResult'>({'newKey': False, '_children': [], 'createdIndex': 1662, 'modifiedIndex': 1662, 'value': u'{"ipv4_gateway": "0.0.0.1", "name": "tapbe12be96-69", "state": "active", "ipv6_nets": [], "mac": "00:01:02:03:04:05", "ipv4_nets": ["10.112.0.210/32"], "profile_ids": ["securitygroup3"]}', 'expiration': None, 'key': u'/calico/v1/host/felixhost-7/workload/openstack/be12be96-69bb-484c-a716-5ddab1f77111/endpoint/be12be96-69bb-484c-a716-5ddab1f77111', 'ttl': None, 'action': None, 'dir': False}), 'raft_index': 2977, '_children': [], 'createdIndex': 2291, 'modifiedIndex': 2291, 'value': u'{"ipv4_gateway": "0.0.0.1", "name": "tapbe12be96-69", "state": "active", "ipv6_nets": [], "mac": "00:01:02:03:04:05", "ipv4_nets": ["10.112.0.210/32"], "profile_ids": ["securitygroup3"]}', 'etcd_index': 2316, 'expiration': None, 'key': u'/calico/v1/host/felixhost-7/workload/openstack/be12be96-69bb-484c-a716-5ddab1f77111/endpoint/be12be96-69bb-484c-a716-5ddab1f77111', 'ttl': None, 'action': u'set', 'dir': False})
110945:2015-06-23 11:08:22,107 [DEBUG][39638/3] calico.etcdutils 45: etcd event set for key /calico/v1/host/felixhost-7/workload/openstack/be12be96-69bb-484c-a716-5ddab1f77111/endpoint/be12be96-69bb-484c-a716-5ddab1f77111
110946:2015-06-23 11:08:22,107 [DEBUG][39638/3] calico.etcdutils 64: Found handler for event set for /calico/v1/host/felixhost-7/workload/openstack/be12be96-69bb-484c-a716-5ddab1f77111/endpoint/be12be96-69bb-484c-a716-5ddab1f77111, captures: {'hostname': u'felixhost-7', 'orchestrator': u'openstack', 'endpoint_id': u'be12be96-69bb-484c-a716-5ddab1f77111', 'workload_id': u'be12be96-69bb-484c-a716-5ddab1f77111'}
110947:2015-06-23 11:08:22,107 [DEBUG][39638/3] calico.felix.fetcd 405: Endpoint EndpointId<be12be96-69bb-484c-a716-5ddab1f77111> updated
110948:2015-06-23 11:08:22,108 [DEBUG][39638/3] calico.felix.fetcd 600: Validated endpoint : {'ipv4_gateway': '0.0.0.1', 'name': u'tapbe12be96-69', 'state': 'active', 'ipv6_nets': [], 'mac': '0:1:2:3:4:5', 'ipv4_nets': ['10.112.0.210/32'], 'profile_ids': ['securitygroup3']}
110949:2015-06-23 11:08:22,108 [DEBUG][39638/3] calico.felix.actor 479: Message 3881f9115b3c (on_endpoint_update) sent by EtcdWatcher (processing d16dab847786) to UpdateSplitter, queue length 0
110950:2015-06-23 11:08:22,108 [DEBUG][39638/3] calico.felix.fetcd 326: About to wait for etcd update 2292
110951:2015-06-23 11:08:22,109 [DEBUG][39638/8] calico.felix.actor 224: Message 3881f9115b3c (on_endpoint_update) recd by UpdateSplitter from EtcdWatcher (processing d16dab847786), queue length 0
110952:2015-06-23 11:08:22,109 [INFO][39638/8] calico.felix.splitter 159: Endpoint update for EndpointId<be12be96-69bb-484c-a716-5ddab1f77111>.
110953:2015-06-23 11:08:22,109 [DEBUG][39638/8] calico.felix.actor 479: Message 99ea27888a3a (on_endpoint_update) sent by UpdateSplitter (processing 3881f9115b3c) to IpsetManager(IPv4), queue length 0
110954:2015-06-23 11:08:22,109 [DEBUG][39638/8] calico.felix.actor 479: Message fecf5ab6d9c8 (on_endpoint_update) sent by UpdateSplitter (processing 3881f9115b3c) to IpsetManager(IPv6), queue length 0
110955:2015-06-23 11:08:22,109 [DEBUG][39638/8] calico.felix.actor 479: Message a41b8012ab88 (on_endpoint_update) sent by UpdateSplitter (processing 3881f9115b3c) to EndpointManager(IPv4), queue length 0
110956:2015-06-23 11:08:22,110 [DEBUG][39638/8] calico.felix.actor 479: Message cacd1ce3bc0f (on_endpoint_update) sent by UpdateSplitter (processing 3881f9115b3c) to EndpointManager(IPv6), queue length 0
110957:2015-06-23 11:08:22,110 [DEBUG][39638/8] calico.felix.actor 245: Finishing message batch
110958:2015-06-23 11:08:22,110 [DEBUG][39638/11] calico.felix.actor 224: Message 99ea27888a3a (on_endpoint_update) recd by IpsetManager(IPv4) from UpdateSplitter (processing 3881f9115b3c), queue length 0
110959:2015-06-23 11:08:22,110 [DEBUG][39638/11] calico.felix.ipsets 304: Endpoint EndpointId<be12be96-69bb-484c-a716-5ddab1f77111> updated
110960:2015-06-23 11:08:22,110 [DEBUG][39638/11] calico.felix.actor 245: Finishing message batch
110961:2015-06-23 11:08:22,110 [INFO][39638/11] calico.felix.ipsets 424: Finishing batch, sending updates to any dirty tags..
110962:2015-06-23 11:08:22,110 [INFO][39638/11] calico.felix.ipsets 426: Finished sending updates to dirty tags.
```

If I grep through the log for similar occurrences, and plot the frequency against time:

![download](https://cloud.githubusercontent.com/assets/301220/8305716/bf445048-19a9-11e5-9b0f-b455ff3696f6.png)

Lots of logs like that at 60s intervals. Possibly I'm barking up the wrong tree entirely, but that seems significant to me.

**Comment by alexwlchan** [comment]
I'm seeing similar logs in the plugin log for the test framework:

    Write port {{port info}} to etcd

500 logs at a time, once every 60s, one for every endpoint. I'm going to sprinkle some exception logging through the codebase, so that I can get some tracebacks and work out exactly where the resync is coming from.

**Comment by Lukasa** [comment]
That resync is coming from the resync, which is a 60s timer. What matters is: why does the plugin think all the ports are missing?

**Comment by alexwlchan** [comment]
Turns out this was a bug in our simulated scale test framework, not Felix. As part of the resync, mech_calico gets a list of missing ports with `get_ports()` from the Neutron plugin. That gets stubbed out in the test framework as follows:

```python
def get_ports(self, context, filters):
    return self.driver.ports.values()
```

Although mech_calico was passing a filter that shouldn't match anything (`{'id': set([])}`) to get a list of missing ports, that was being ignored. Hence every port gets treated as missing. Oops.

I've tweaked the stub so that it returns an empty list if it detects this empty resync filter. This seems to have fixed the issue:

* etcd disk usage still increases slowly, at ~4KB every 10 seconds, but it's much more gradual than before.
* CPU usage for both etcd and calico-felix is way down.
* etcd's (resident) memory usage is noticeably down.
* Once the network stabilises, calico-felix's CPU usage drops to 0.

This explains why we couldn't repro the issue on a vanilla OpenStack setup.
